File: examples/typegraphs/programmable-api-gateway.py
# skip:start
import yaml

from typegraph import typegraph, Policy, t, Graph
from typegraph.graph.params import Cors
from typegraph.runtimes.deno import DenoRuntime

# skip:end


@typegraph(
    # skip:next-line
    cors=Cors(allow_origin=["https://metatype.dev", "http://localhost:3000"]),
)
def programmable_api_gateway(g: Graph):
    deno = DenoRuntime()

    public = Policy.public()
    roulette_access = deno.policy("roulette", "() => Math.random() < 0.5")

    my_api_format = """
    static_a:
        access: roulette_access
        foo: rab
    static_b:
        access: public
        foo: bar
    """

    for field, static_vals in yaml.safe_load(my_api_format).items():
        g.expose(
            public if static_vals.pop("access") == "public" else roulette_access,
            **{field: deno.static(t.struct({"foo": t.string()}), static_vals)},
        )

File: examples/typegraphs/programmable-api-gateway.ts
// skip:start

import { Policy, t, typegraph } from "@typegraph/sdk/index.ts";
import { DenoRuntime } from "@typegraph/sdk/runtimes/deno.ts";

// skip:end

typegraph(
  {
    name: "programmable-api-gateway",
    // skip:next-line
    cors: { allowOrigin: ["https://metatype.dev", "http://localhost:3000"] },
  },
  (g) => {
    const deno = new DenoRuntime();

    const pub = Policy.public();
    const roulette_access = deno.policy(
      "roulette",
      "() => Math.random() < 0.5"
    );

    // skip:next-line
    const myApiFormat = {
      static_a: { foo: "rab", access: "roulette_access" },
      static_b: { foo: "bar", access: "public" },
    };

    for (const [k, static_vals] of Object.entries(myApiFormat)) {
      const policy = static_vals["access"] == "public" ? pub : roulette_access;
      g.expose(
        {
          [k]: deno.static(t.struct({ foo: t.string() }), {
            foo: static_vals["foo"],
          }),
        },
        policy
      );
    }
  }
);
